src/System.Management.Automation/engine/parser/tokenizer.cs

```diff
@@ -3294,15 +3329,46 @@ private static bool TryGetNumberValue(string strNum, bool hex, bool real, char s
                     style = hex ? NumberStyles.AllowHexSpecifier : NumberStyles.AllowLeadingSign;
 
                     long longValue;
-                    if (suffix == 'l' || suffix == 'L')
+                    switch (suffix)
                     {
-                        if (long.TryParse(strNum, style, NumberFormatInfo.InvariantInfo, out longValue))
-                        {
-                            result = longValue * multiplier;
-                            return true;
-                        }
-                        result = null;
-                        return false;
+                        case NumberSuffixFlags.Long:
+                            if (long.TryParse(strNum, style, NumberFormatInfo.InvariantInfo, out longValue))
+                            {
+                                result = longValue * multiplier;
+                                return true;
+                            }
+
+                            result = null;
+                            return false;
+                        case NumberSuffixFlags.Short:
+                            if (short.TryParse(strNum, style, NumberFormatInfo.InvariantInfo, out short s))
+                            {
+                                result = (short)(s * multiplier);
+                                return true;
+                            }
+
+                            result = null;
+                            return false;
+                        default:
+                            if (suffix.HasFlag(NumberSuffixFlags.Unsigned) && ulong.TryParse(strNum, style, NumberFormatInfo.InvariantInfo, out ulong u))
+                            {
+                                u *= (ulong)multiplier;
+                                if (suffix.HasFlag(NumberSuffixFlags.Short) && u <= ushort.MaxValue)
+                                {
+                                    result = (ushort)u;
+                                }
+                                else if (suffix.HasFlag(NumberSuffixFlags.Long) || u > uint.MaxValue)
+                                {
+                                    result = u;
+                                }
+                                else
+                                {
+                                    result = (uint)u;
+                                }
```

Comment:
Maybe more clear:
```c#
                               else if (suffix.HasFlag(NumberSuffixFlags.Long) || u <= uint.MaxValue)  
                                {  
                                     result = (uint)u;
                                 }  
                                else  
                                {  
                                    // ulong
                                    result = u;
                                 }  
```

Reply:
Oh, you mean have the ulong case come last and alter the penultimate else if... Hmm. Yeah, OK, I can see the merit in that. Will do.

Reply:
Closed.